base/expr.jl

```diff
@@ -184,6 +184,57 @@ macro noinline(ex)
     esc(isa(ex, Expr) ? pushmeta!(ex, :noinline) : ex)
 end
 
+"""
+    @pure
+
+Annotates a function to ease type inference by strictly specifying the
+output is completely determined by the input.
+
+Note: Misuse can lead to regressions.
+
+Do not use if the function involved:
+
+- Involves globals, pointers
+- Recurses
```

Comment:
What’s wrong with recursion?

Reply:
It came up during a Discourse post and @mbauman mentioned the restriction.
https://discourse.julialang.org/t/pure-macro/3871

Reply:
typejoin is pure and does recursion - seems to be doing OK

Reply:
Thanks! Will ammend.